iocore/net/quic/test/test_QUICAckFrameCreator.cc

```diff
@@ -64,6 +64,80 @@ TEST_CASE("QUICAckFrameCreator", "[quic]")
   CHECK(frame != nullptr);
   CHECK(frame->num_blocks() == 1);
   CHECK(frame->largest_acknowledged() == 10);
+  CHECK(frame->ack_block_section()->first_ack_block_length() == 1);
+  CHECK(frame->ack_block_section()->begin()->gap() == 2);
+}
+
+TEST_CASE("QUICAckFrameCreator_loss_recover", "[quic]")
+{
+  QUICAckFrameCreator creator;
+  std::unique_ptr<QUICAckFrame, QUICFrameDeleterFunc> frame = {nullptr, nullptr};
+
+  // Initial state
+  frame = creator.create();
+  CHECK(frame == nullptr);
+
+  creator.update(2, true);
+  creator.update(5, true);
+  creator.update(6, true);
+  creator.update(8, true);
+  creator.update(9, true);
+
+  frame = creator.create();
+  CHECK(frame != nullptr);
+  CHECK(frame->num_blocks() == 2);
+  CHECK(frame->largest_acknowledged() == 9);
   CHECK(frame->ack_block_section()->first_ack_block_length() == 2);
```

Comment:
I'm not sure here, why dose the ack_block_length is 2.

Reply:
Because it's for packet number 6 and 7, if I understand the spec correctly.

Reply:
The spec says the first_ack_block_length is the `number of contiguous packets preceding the Largest Acknowledged`.

We only received 10 here.

```
A variable-length integer indicating the number of contiguous packets preceding the Largest Acknowledged that are being acknowledged.
```

Reply:
I see.  I thought opposite (a block for 6 and 7 follows a block for 10) but probably you're right.

So, the sequence of ack blocks here has to be:
First block: for 10
Second block (the block after the first gap): for 7 and 6
Right? Then `first_ack_block_length` should be 1.

Reply:
Is my understanding correct now?

I'm +1 on the change to introduce `QUICAckPacketNumbers` but I want to confirm that we are on the same.

Reply:
I'm implementing ack block section parser (it's been left as TODO surprisingly) and I confirmed that your understanding is right. I'll approve this when CI jobs complete.